controller/object_store_controller.go

```diff
@@ -0,0 +1,1170 @@
+								Backend:  s3backend,
+							},
+						},
+					},
+				},
+			}
+
+			ingress := &networkingv1.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            name,
+					Namespace:       osc.namespace,
+					Labels:          types.GetObjectStoreLabels(store),
+					OwnerReferences: osc.ds.GetOwnerReferencesForObjectStore(store),
+				},
+				Spec: networkingv1.IngressSpec{
+					Rules: []networkingv1.IngressRule{
+						baserule,
+						wildcardrule,
+					},
+				},
+			}
+
+			if endpoint.TLS.Name != "" {
+				ingress.Spec.TLS = []networkingv1.IngressTLS{
+					{
+						SecretName: endpoint.TLS.Name,
+						Hosts: []string{
+							endpoint.DomainName,
+							fmt.Sprintf("*.%v", endpoint.DomainName),
+						},
+					},
+				}
+			}
+
+			_, err := osc.ds.CreateIngress(osc.namespace, ingress)
+			if err != nil && !datastore.ErrorIsAlreadyExists(err) {
+				store.Status.State = longhorn.ObjectStoreStateError
+				return []*networkingv1.Ingress{}, store, err
+			}
+
+			store.Status.Endpoints = append(store.Status.Endpoints, endpoint.DomainName)
+			ingresses = append(ingresses, ingress)
+		} else {
+			// if there was an api error
+			return []*networkingv1.Ingress{}, store, err
+		}
+	}
+
+	return ingresses, store, nil
+}
+
+func (osc *ObjectStoreController) createVolume(
+	store *longhorn.ObjectStore,
+	pvc *corev1.PersistentVolumeClaim,
+) (*longhorn.Volume, error) {
+	vol := longhorn.Volume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      genPVName(store),
+			Namespace: osc.namespace,
+			Labels:    types.GetObjectStoreLabels(store),
+			Annotations: map[string]string{
+				types.LonghornAnnotationObjectStoreName: store.Name,
+			},
+			OwnerReferences: osc.ds.GetOwnerReferencesForPVC(pvc),
+		},
+		Spec: longhorn.VolumeSpec{
+			Size:                        resourceAsInt64(store.Spec.Storage.Size),
+			Frontend:                    longhorn.VolumeFrontendBlockDev,
+			AccessMode:                  longhorn.AccessModeReadWriteOnce,
+			NumberOfReplicas:            store.Spec.Storage.NumberOfReplicas,
+			ReplicaSoftAntiAffinity:     store.Spec.Storage.ReplicaSoftAntiAffinity,
+			ReplicaZoneSoftAntiAffinity: store.Spec.Storage.ReplicaZoneSoftAntiAffinity,
+			ReplicaDiskSoftAntiAffinity: store.Spec.Storage.ReplicaDiskSoftAntiAffinity,
+			DiskSelector:                store.Spec.Storage.DiskSelector,
+			NodeSelector:                store.Spec.Storage.NodeSelector,
+			DataLocality:                store.Spec.Storage.DataLocality,
+			FromBackup:                  store.Spec.Storage.FromBackup,
+			StaleReplicaTimeout:         store.Spec.Storage.StaleReplicaTimeout,
+			ReplicaAutoBalance:          store.Spec.Storage.ReplicaAutoBalance,
+			RevisionCounterDisabled:     store.Spec.Storage.RevisionCounterDisabled,
+			UnmapMarkSnapChainRemoved:   store.Spec.Storage.UnmapMarkSnapChainRemoved,
+			BackendStoreDriver:          store.Spec.Storage.BackendStoreDriver,
```

Comment:
Great question!

There are 2 approaches that I am thinking about:

### approach 1:
* We will not sync them after the volume is created. The value in `ObjectStore.Spec.Storage` is used when provisioning the volume the first time only. This is similar to the parameters inside the PV when the volume is dynamically provisioned. 
* There is an exception though, the size should be synced from ObjectStore to volume/PV/PVC. similar to how users can change the size of PVC and get sync to Longhorn volume.
* Modification to all other fields of `ObjectStore.Spec.Storage` is blocked from webhook after ObjectStore is created

### approach 2:
* We will sync the fields inside `ObjectStore.Spec.Storage` to volume/PV/PVC.
* This is one-way sync, the source of truth is the fields in `ObjectStore.Spec.Storage`
* The volume is fully controlled by ObjectStore so users would need to change the `ObjectStore.Spec.Storage` for these fields. 
* If users change the volume for the fields that are controlled by `ObjectStore.Spec.Storage`, the ObjectStore will revert it immediately. Again, users would need to change `ObjectStore.Spec.Storage` for these fields because ObjectStore is fully controlling the volume.

I am more inclined to the approach 2 because:
* The approach 1 sometimes leads to the situation that the value inside volume.Spec is different than the value inside ObjectStore.Spec.Storage which is a bit uncomfortable 
* It feels a bit award that users cannot change the `ObjectStore.Spec.Storage` fields.

Reply:
I prefer approach 1 as well. Making `ObjectStore.Spec.Storage` immutable after ObjectStore creation is acceptable. Besides, can we put field `size` outside of `ObjectStore.Spec.Storage`? Otherwise, users may be confused as to why some fields in `ObjectStore.Spec.Storage` are editable while others are not

Reply:
Ok. Thanks both. I can handle this in a separate PR which has some other improvements that is coming up soon. There are some aspects which are hard to convey through PR review comments. To avoid the delaying of this PR and increase the efficiency, let's move this PR forward first. Is it ok?